keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,886 @@
+  existing Kubernetes API objects, and would confuse any code that
+  tries to use `netutils.ParseIPSloppy`, so we need to be careful to
+  not accept them.
+
+We will update all in-tree API types to use appropriate
+`utilvalidation` methods for IP/CIDR validation, and update the
+validation functions to allow only unambiguous values.
+
+The fields whose validation will be updated are:
+
+  - in `core`:
+    - `endpoints.subsets[].addresses[].ip`
+    - `endpoints.subsets[].notReadyAddresses[].ip`
+    - `node.spec.podCIDRs[]`
+    - `pod.spec.dnsConfig.nameservers[]`
+    - `pod.spec.hostAliases[].ip`
+    - `pod.spec.hostIP`
+    - `pod.spec.hostIPs[]`
+    - `pod.spec.podIP`
+    - `pod.spec.podIPs[]`
+    - `service.spec.clusterIP`
+    - `service.spec.clusterIPs[]`
+    - `service.spec.externalIPs[]`
+    - `service.spec.loadBalancerSourceRanges[]`
+    - `service.status.loadBalancer.ingress[].ip`
+
+  - in `networking`:
+    - `ingress.status.loadBalancer.ingress[].ip`
+    - `networkpolicy.spec.egress[].to[].ipBlock.cidr`
+    - `networkpolicy.spec.egress[].to[].ipBlock.except[]`
+    - `networkpolicy.spec.ingress[].from[].ipBlock.cidr`
+    - `networkpolicy.spec.ingress[].from[].ipBlock.except[]`
+    - `serviceCIDR.spec.cidrs[]`
+
+  - in `discovery`:
+    - `endpointslice.endpoints[].addresses[]`
+
+#### Updates to pre-existing objects with pre-existing invalid fields
+
+We want to allow making changes to objects containing invalid
+IPs/CIDRs, as long as the IPs/CIDRs themselves are not changed.
+However, in the case of array-valued fields, it seems best to allow
+adding and removing new valid IPs/CIDRs without needing to
+simultaneously fix pre-existing invalid IPs/CIDRs.
+
+So, the rule will be rougly: "When validating an Update, any IP/CIDR
+that exists in the old version of the field is allowed to exist in the
+new version of the field." For NetworkPolicy, that is extended to "any
+CIDR that exists anywhere in the old version of the object is allowed
+to exist anywhere in the new version of the object" (to allow
+inserting new rules into a policy without having to fix invalid CIDRs
+in unrelated later rules).
+
+However, `Endpoints` and `EndpointSlice` will be treated differently,
+because (a) they are large enough that doing additional validation on
+them could be noticeably slow, (b) in most cases, they are generated
+by controllers that only write out valid IPs anyway, (c) in the case
+of `EndpointSlice`, if we were going to allow moving bad IPs around
+within a slice without revalidation, then we ought to allow moving
+them between related slices too, which we can't really do.
+
+So for `Endpoints` and `EndpointSlice`, the rule will be that invalid
+IPs are only allowed to remain unfixed if the update leaves the entire
+`.subsets` / `.addresses` unchanged. So you can edit the labels or
+annotations without fixing invalid endpoint IPs, but you can't add a
+new IP while leaving existing invalid IPs in place.
```

Comment:
EndpointSlices list of Endpoints is atomic
```
	// endpoints is a list of unique endpoints in this slice. Each slice may
	// include a maximum of 1000 endpoints.
	// +listType=atomic
	Endpoints []Endpoint `json:"endpoints" protobuf:"bytes,2,rep,name=endpoints"`
```

and Endpoints subsets too
```
	// The set of all endpoints is the union of all subsets. Addresses are placed into
	// subsets according to the IPs they share. A single address with multiple ports,
	// some of which are ready and some of which are not (because they come from
	// different containers) will result in the address being displayed in different
	// subsets for the different ports. No address will appear in both Addresses and
	// NotReadyAddresses in the same subset.
	// Sets of addresses and ports that comprise a service.
	// +optional
	// +listType=atomic
	Subsets []EndpointSubset `json:"subsets,omitempty" protobuf:"bytes,2,rep,name=subsets"`
```

Reply:
It's not about the _mechanics_ of the update, it's about the _result_. The validation code can't tell whether you did an Update or a Patch anyway.

eg, if you have a service with `externalIPs: ["001.002.003.004"]`, and your controller tries to reconcile it to `["001.002.003.004", "5.6.7.8"]`, then the proposed validation rule will allow it. Regardless of whether you actually did it with a Patch or not, the change is semantically interpreted as as "the controller appended a new value and the old value didn't change", and the validation allows that for `externalIPs`.

But if you have an EndpointSlice with `endpoints: [ { addresses: ["001.002.003.004"] } ]` and your controller tries to reconcile it to `endpoints: [ { addresses: ["001.002.003.004"] }, { addresses: ["5.6.7.8"] } ]`, the validation will _not_ allow it, regardless of whether you did an Update or a Patch, because the proposed validation rule for Endpoints/EndpointSlice doesn't have that "and the old value didn't change" exception, for efficiency.

Reply:
it looks simpler to avoid these problems, if the update of the field has to be atomic the it must pass the new validation

Reply:
The question isn't whether the update is an atomic replacement or an incremental patch _from the apiserver's perspective_, it's whether the update is an atomic replacement or an incremental patch _from the client's perspective_.

That is, if a Service has `externalIPs: ["001.002.003.004"]`, and a client does:

    svc, _ := client.CoreV1().Services(ns).Get(ctx, name, getOptions)
    svc.Spec.ExternalIPs = append(svc.Spec.ExternalIPs, "5.6.7.8")
    _, err := client.CoreV1().Services(ns).Update(ctx, update, updateOptions)

then does that validate or not?

(And the KEP currently says, for the case of everything except Endpoints and EndpointSlice, yes, you can do that, and it will still validate. But for Endpoints and EndpointSlice you can't, because it would be annoying to make it work and it's probably not needed.)

Reply:
(And maybe we _don't_ want to do it that way, I'm just trying to clarify what the issue is.)

Reply:
So first off, that's exactly what I'm proposing for Endpoints and EndpointSlice; if `endpoints.subsets` or `endpointslice.endpoints` changes at all, it must be fully valid according to the new rules.

But what about array-valued fields which are explicitly _not_ `listType=atomic`? `pod.spec.podIPs` is type `map`, and `node.spec.podCIDRs` is type `set`. So should you be allowed to modify one element of those arrays without touching the other (possibly-invalid) elements?

> so we can be completely sure at GA that removing the old parsers is safe

("the old parsers" (the sloppy parsers in netutils) don't get removed as part of this KEP; they just stop being the sole definition of "valid".)

Reply:
>  if endpoints.subsets or endpointslice.endpoints changes at all, it must be fully valid according to the new rules.

yeah, just applying the same logic everywhere, I'm specially worried about Services.

> pod.status.podIPs

are those not normalized somewhere across the kubelet path? besides they are not inmutable, are only used by the kubelet and never updated by the kubelet, so this does not look a risk

> node.spec.podCIDRs

this is inmutable, can not be update

> ("the old parsers" (the sloppy parsers in netutils) don't get removed as part of this KEP; they just stop being the sole definition of "valid".)

yeah, my bad, is just "put those parser to oblivion"